### PR TITLE
Mamdani inference system

### PR DESCRIPTION
**Pull request by Adam-Chlopowiec**

### CODE REVIEW

File: doggos/inference/mamdani_inference_system.py [hunk 1]
```diff
@@ -1,6 +1,112 @@
...
+
+    def __validate_consequents(self):
+        for rule in self._rule_base:
+            if not isinstance(rule.consequent, MamdaniConsequent):
+                raise ValueError("All rule consequents must be mamdani consequents")
+
+    def __is_consequent_type1(self):
+        return isinstance(self.__rule_base[0].consequent.clause.fuzzy_set, Type1FuzzySet)
+
+    def __get_domain_and_memberships_for_it2(self, features: Dict[Clause, List[MembershipDegree]]) \
```
Comment (Adrian-Chlopowiec) [inline doggos/inference/mamdani_inference_system.py:85]: Consider name:
__get_domain_and_outputs_for_it2

File: doggos/inference/mamdani_inference_system.py [hunk 2]
```diff
@@ -1,6 +1,112 @@
...
+        Extracts domain and membership functions from rule base
+        :param features: dictionary of linguistic variables and their values
+        :return: domain, lower membership functions and upper membership functions extracted from rule base
+        """
+        domain, membership_functions = self.__get_domain_and_membership_functions(features)
+        lmfs = [output[0] for output in membership_functions]
+        umfs = [output[1] for output in membership_functions]
+        return domain, lmfs, umfs
+
+    def __get_domain_and_membership_functions(self, features: Dict[Clause, List[MembershipDegree]]) \
```
Comment (Adrian-Chlopowiec) [inline doggos/inference/mamdani_inference_system.py:97]: Consider name:
__get_domain_and_outputs

File: doggos/inference/mamdani_inference_system.py (hunk 1 above)
Comment (Adam-Chlopowiec) [reply doggos/inference/mamdani_inference_system.py:94]: Would rather change output variables inside of functions to membership_function - provides better description

File: doggos/inference/defuzzification_algorithms.py [hunk 3]
```diff
@@ -1,31 +1,38 @@
+from collections import Iterable
```
Comment (konrad-karanowski) [inline doggos/inference/defuzzification_algorithms.py:1]: from collections.abc import Iterable

File: doggos/inference/defuzzification_algorithms.py [hunk 4]
```diff
@@ -1,31 +1,38 @@
...
-    return domain[np.where(cut == maximum)[0]]
+    return domain[np.where(cut == maximum)[0][0]]
 
 
-def middle_of_maximum(domain, cut):
+def middle_of_maximum(domain, membership_functions):
+    cut = __membership_func_union(membership_functions)
     maximum = np.max(cut)
     indices = np.where(cut == maximum)[0]
     size = len(indices)
```
Comment (konrad-karanowski) [inline doggos/inference/defuzzification_algorithms.py:29]: If "indicies" is a vector use size

File: doggos/inference/defuzzification_algorithms.py [hunk 5]
```diff
@@ -1,31 +1,38 @@
...
     middle = int(size / 2)
     return domain[[indices[middle]]]
 
 
-def mean_of_maxima(domain, cut):
+def mean_of_maxima(domain, membership_functions):
+    cut = __membership_func_union(membership_functions)
     maximum = np.max(cut)
     indices = np.where(cut == maximum)[0]
     size = len(indices)
```
Comment (konrad-karanowski) [inline doggos/inference/defuzzification_algorithms.py:38]:
```suggestion
    size = indicies.size
```

File: tests/test_inference/test_mamdani_inference_system.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        output = mamdani.infer(method, {})
+        assert output == approx(0.5)
+
+    @pytest.mark.parametrize('method, expected', zip([largest_of_maximum, middle_of_maximum, smallest_of_maximum,
+                                                      mean_of_maxima], [0.75, 0.5, 0.25, 0.5]))
+    def test_maxima(self, patch_get_degrees, patch_consequent_type1_true,
+                    patch_domain_and_membership_functions_for_maxima, method, expected):
+        mamdani = MamdaniInferenceSystem([])
+        output = mamdani.infer(method, {})
+        assert output == approx(expected)
```
Comment (konrad-karanowski) [inline tests/test_inference/test_mamdani_inference_system.py:81]: All tests fail

**Review by konrad-karanowski — CHANGES_REQUESTED** [review]

**Review by Adrian-Chlopowiec — APPROVED** [review]

**Review by konrad-karanowski — APPROVED** [review]

**Pull request merged**